engine/evd.ml

```diff
@@ -584,6 +584,16 @@ let undefined_map d = d.undf_evars
 
 let drop_all_defined d = { d with defn_evars = EvMap.empty }
 
+let reset_non_goal_defined d =
+  let f evk info =
+    match snd info.evar_source with
+    (* These are the two evar kinds used for existing goals *)
+    (* see Proofview.mark_in_evm *)
+    | Evar_kinds.GoalEvar | Evar_kinds.VarInstance _ -> info
+    | _ -> { info with evar_body = Evar_empty } in
+  let defn_evars = EvMap.Smart.mapi f d.defn_evars in
```

Comment:
This is going to be **horribly** expensive. You can't expect your basic functions to be `O(n)` on the number of *defined* evars.

Reply:
OK, right.

What do you propose then? The objective is to type a pattern (inserting e.g. coercions) w/o losing the original holes of the pattern. It might be useful also for typing notations.

Reply:
Indeed, but there are several problems tacked on this. IIRC the main problem was that some evars that needed to be considered as pattern holes were automagically solved by the constraint solver of the pretyper. If anything, I'd rather provide a way to prevent these evars to get modified against the will of the user.

Reply:
If "automagically solved" means physically modified, this would indeed defeat the algorithm.

Maybe could we prepare a problem of the form `let id1:= _ in ... let idn := _ in pattern`, filling the holes with variables, and replace the `id`s by hole again after typing. This is another hack but which only assume that the structure of let-in is preserved.

More generally, there can in principle be more "complex" metavariables in patterns than what pretyping is able to deal with (second-order metavariables, metavariables for pattern-matching, metavariables of recursive patterns, ...). So maybe a proper type-checker for patterns would be more reasonable, even if it would duplicate some code.

Reply:
> So maybe a proper type-checker for patterns would be more reasonable, even if it would duplicate some code.

That was one of the arguments for an open-recursion style pretyper implementation :disappointed:.

Reply:
I did not follow all discussions. Is the idea of an open-recursion style pretyper abandoned?